politeiawww/dcc.go

```diff
@@ -538,3 +574,273 @@ func (p *politeiawww) processGetDCCs(gds cms.GetDCCs) (*cms.GetDCCsReply, error)
+			ErrorCode: www.ErrorStatusUserActionNotAllowed,
+		}
+	}
+
+	// Check to make sure that the DCC is still active
+	if dcc.Status != cms.DCCStatusActive {
+		return nil, www.UserError{
+			ErrorCode: cms.ErrorStatusCannotCommentOnDCC,
+		}
+	}
+
+	// Create the change record.
+	c := backendDCCSupportOppositionMetadata{
+		Version:   backendDCCSupposeOppositionMetadataVersion,
+		PublicKey: u.PublicKey(),
+		Timestamp: time.Now().Unix(),
+		Vote:      sd.Vote,
+	}
+	blob, err := encodeBackendDCCSupportOppositionMetadata(c)
+	if err != nil {
+		return nil, err
+	}
+
+	challenge, err := util.Random(pd.ChallengeSize)
+	if err != nil {
+		return nil, err
+	}
+
+	pdCommand := pd.UpdateVettedMetadata{
+		Challenge: hex.EncodeToString(challenge),
+		Token:     sd.Token,
+		MDAppend: []pd.MetadataStream{
+			{
+				ID:      mdStreamDCCSupportOpposition,
+				Payload: string(blob),
+			},
+		},
+	}
+
+	responseBody, err := p.makeRequest(http.MethodPost, pd.UpdateVettedMetadataRoute, pdCommand)
+	if err != nil {
+		return nil, err
+	}
+
+	var pdReply pd.UpdateVettedMetadataReply
+	err = json.Unmarshal(responseBody, &pdReply)
+	if err != nil {
+		return nil, fmt.Errorf("Could not unmarshal UpdateVettedMetadataReply: %v",
+			err)
+	}
+
+	// Verify the UpdateVettedMetadata challenge.
+	err = util.VerifyChallenge(p.cfg.Identity, challenge, pdReply.Response)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cms.SupportOpposeDCCReply{}, nil
+}
+
+func stringInSlice(arr []string, str string) bool {
+	for _, s := range arr {
+		if str == s {
+			return true
+		}
+	}
+
+	return false
+}
+
+// processNewCommentDCC sends a new comment decred plugin command to politeaid
+// then fetches the new comment from the cache and returns it.
+func (p *politeiawww) processNewCommentDCC(nc www.NewComment, u *user.User) (*www.NewCommentReply, error) {
+	log.Tracef("processNewCommentDCC: %v %v", nc.Token, u.ID)
+
+	dcc, err := p.getDCC(nc.Token)
+	if err != nil {
+		if err == cache.ErrRecordNotFound {
+			err = www.UserError{
+				ErrorCode: cms.ErrorStatusDCCNotFound,
+			}
+		}
+		return nil, err
+	}
+
+	// Ensure the public key is the user's active key
+	if nc.PublicKey != u.PublicKey() {
+		return nil, www.UserError{
+			ErrorCode: www.ErrorStatusInvalidSigningKey,
+		}
+	}
+
+	// Validate signature
+	msg := nc.Token + nc.ParentID + nc.Comment
+	err = validateSignature(nc.PublicKey, nc.Signature, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Don't allow comments of just "aye" or "nay" that would be confused
+	// with support or opposition.
+	if nc.Comment == supportString || nc.Comment == opposeString {
+		return nil, www.UserError{
+			ErrorCode: www.ErrorStatusUserActionNotAllowed,
+		}
+	}
+
+	// Validate comment
+	err = validateComment(nc)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check to make sure that dcc isn't already approved.
+	if dcc.Status != cms.DCCStatusActive {
+		return nil, www.UserError{
+			ErrorCode: cms.ErrorStatusCannotCommentOnDCC,
```

Comment:
This user error needs to tell the user why they can't comment on the DCC.  A general "cannot comment" is not helpful.